### PR TITLE
Evaluate and Transfer Plotting

### PR DESCRIPTION
**Pull request by RussellSB**
# Overview
- Added evaluate.py - purpose to run on a test set without weight updates, and save out.pickle of conversions when done
- Added plotting functionality to both evaluate.py and train.py
- Updated Encoder model definition, removed final_block (shared_block) param
- Updated preprocessing to be with defaultdict(list). Just for less verbosity, and consistency with what I did in evaluate.py for feats dictionary

## Plotting Notes
- In train.py it plots the first batch every epoch or few epochs (specified by plot_interval arg)
- In evaluate.py plot_interval arg controls how often it plots a batch (with respect to every few batches, since no epochs in evaluate)
- The code looks kind of repeated for plotting in utils.py. But figured its more readable this way as opposed to having a single method with lots of conditions. Also, need to separate eval from train plot functions, since train is also concerned with plotting the target used in updating weight - while eval isn't.

### Train Plot
image

### Eval Plot
image

### CODE REVIEW

**Comment by RussellSB** [comment]
## Train Discussion
Is it worth mirroring how evaluate.py saves a feats dict of converted output but also in train.py (on its last epoch)? I'm unsure of this. My reservations lie in that if the user wants to get the converted output of one epoch of training batches, they can use evaluate.py on the train set. Only reservation is out.pickle would be saved in out_eval not out_train.

**Comment by RussellSB** [comment]
## Preprocessing Discussion
In evaluate.py for the feats dict, I named the keys with respect to their conversion A2B, B2A. Seems to be a convention in style transfer code to use 'A2B' notation. Should we consider maybe changing the convention of how spkr directories are named in the database? Instead, if we were to name them spkr_A and spkr_B it could be more consistent. And we can iterate over key strings, instead of key int values in preprocessing.py.

Not only would this make things more consistent with "A2B notation" but say if the user were to name directories "spkr_4" or "spkr_7" (corresponding to ids from flickr_audio), it would string sort and still take spkr_4 as A and spkr_7 as B. We can then have n_spkrs controlled by the number of speaker directories in the dataset. Though for now, we'll note that it only works with two for the time being.

So for preprocessing maybe:
1. Read number spkr dirs in dataset
2. assert num_dirs == 2, 'Currently only two speakers accepted'

File: src/evaluate.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+    # Reconstruct speech
+    recon_X1 = G1(Z1)
+    recon_X2 = G2(Z2)
+
+    # Translate speech
+    fake_X1 = G1(Z2)
+    fake_X2 = G2(Z1)
+
+    # Cycle translation
```
Comment (ebadawy) [inline src/evaluate.py:112]: Looks like you are not using cycle translation anywhere. Could be deleted as it is only used for training.

File: src/train.py [hunk 2]
```diff
@@ -72,11 +77,11 @@
 
 if opt.epoch != 0:
     # Load pretrained models
-    encoder.load_state_dict(torch.load("saved_models/%s/encoder_%d.pth" % (opt.model_name, opt.epoch)))
-    G1.load_state_dict(torch.load("saved_models/%s/G1_%d.pth" % (opt.model_name, opt.epoch)))
-    G2.load_state_dict(torch.load("saved_models/%s/G2_%d.pth" % (opt.model_name, opt.epoch)))
-    D1.load_state_dict(torch.load("saved_models/%s/D1_%d.pth" % (opt.model_name, opt.epoch)))
-    D2.load_state_dict(torch.load("saved_models/%s/D2_%d.pth" % (opt.model_name, opt.epoch)))
+    encoder.load_state_dict(torch.load("saved_models/%s/encoder_%02d.pth" % (opt.model_name, opt.epoch)))
```
Comment (ebadawy) [inline src/train.py:80]: Any reason why using `%02d`? This will limit the number of epochs to be 99 max.

File: src/evaluate.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+    # -------------------------------
+    #  Infer with Encoder and Generators
+    # -------------------------------
+
+    # Get shared latent representation
+    mu1, Z1 = encoder(X1)
+    mu2, Z2 = encoder(X2)
+
+    # Reconstruct speech
```
Comment (ebadawy) [inline src/evaluate.py:104]: If you are only looking for the voice conversion case and not the same speaker reconstruction you can skip this step and use the original sample instead for the comparison. this should make  the run time a bit faster.

File: src/evaluate.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    if i % opt.plot_interval == 0:
+        plot_batch_eval(opt.model_name, 'plot_A2B', i, X1, recon_X1, fake_X2)
+        plot_batch_eval(opt.model_name, 'plot_B2A', i, X2, recon_X2, fake_X1)
+        
+    # Append batch output to features dictionary
+    feats['A2B'].append([spect for spect in to_numpy(fake_X2)])
+    feats['B2A'].append([spect for spect in to_numpy(fake_X1)])
+        
+# Save converted output in pickle format
+pickle.dump(feats,open('out_eval/%s/out.pickle'%(opt.model_name),'wb'))
```
Comment (ebadawy) [inline src/evaluate.py:128]: maybe also save the waveform? you can use `reconstruct_waveform` function in `utils.py` for that.

**Comment by ebadawy** [comment]
> ## Preprocessing Discussion
> In evaluate.py for the feats dict, I named the keys with respect to their conversion A2B, B2A. Seems to be a convention in style transfer code to use 'A2B' notation. Should we consider maybe changing the convention of how spkr directories are named in the database? Instead, if we were to name them spkr_A and spkr_B it could be more consistent. And we can iterate over key strings, instead of key int values in preprocessing.py.
> 
> Not only would this make things more consistent with "A2B notation" but say if the user were to name directories "spkr_4" or "spkr_7" (corresponding to ids from flickr_audio), it would string sort and still take spkr_4 as A and spkr_7 as B. We can then have n_spkrs controlled by the number of speaker directories in the dataset. Though for now, we'll note that it only works with two for the time being.


This seems to be a good idea; however, what do you think about having more than two speakers? we will need to use something else other than A B naming convention in that case. I will leave changing the naming to you if you want to do it or not.

> 
> So for preprocessing maybe:
> 
> 1. Read number spkr dirs in dataset
> 2. assert num_dirs == 2, 'Currently only two speakers accepted'

I like this assertion, we need to have it 

> 
> Would also need to reflect such changes in data_proc.py

File: src/evaluate.py (hunk 4 above)
Comment (ebadawy) [inline src/evaluate.py:128]: probably you will need to mention the epoch number used to generate this output.

**Comment by ebadawy** [comment]
> ## Train Discussion
> Is it worth mirroring how evaluate.py saves a feats dict of converted output but also in train.py (on its last epoch)? I'm unsure of this. My reservations lie in that if the user wants to get the converted output of one epoch of training batches, they can use evaluate.py on the train set. Only reservation is out.pickle would be saved in out_eval not out_train.

having a separated `evaluate.py` seems to be a bit redundant since it uses a lot of the training code. what it be better to include that in `train.py` and have additional argument to how frequent the plots are being generated?

File: src/evaluate.py (hunk 1 above)
Comment (RussellSB) [reply src/evaluate.py:112]: That's a fair suggestion, but I'm using the cyclic reconstruction in the plot function. 

I think it might be good to keep just to have a full visualisation on what it learnt to do. Could help in observing performance changes with architectural modifications - for unseen data

File: src/train.py (hunk 2 above)
Comment (RussellSB) [reply src/train.py:80]: Its just for file name formatting purposes so that in the default version it names files with 00, 01, 02, ...., 99. Its just a small touch so things are easily sorted by name when it comes to loading files from the directory.

If someone were to go over epoch 99, it will still take it as it would previously and will just surpass the formatting constraint. Like 100, 101, ... 100000.

I think it's a neat touch for at least the initial parameters (and think it would be overkill to make digit formatting flexible for any n_epoch) - I could remove the %02d if you prefer.

File: src/evaluate.py (hunk 1 above)
Comment (RussellSB) [reply src/evaluate.py:112]: Agreed!

File: src/evaluate.py (hunk 3 above)
Comment (RussellSB) [reply src/evaluate.py:104]: I'll delete this too

File: src/evaluate.py (hunk 4 above)
Comment (RussellSB) [reply src/evaluate.py:128]: Good idea

**Comment by RussellSB** [comment]
> > ## Train Discussion
> > Is it worth mirroring how evaluate.py saves a feats dict of converted output but also in train.py (on its last epoch)? I'm unsure of this. My reservations lie in that if the user wants to get the converted output of one epoch of training batches, they can use evaluate.py on the train set. Only reservation is out.pickle would be saved in out_eval not out_train.
> 
> having a separated `evaluate.py` seems to be a bit redundant since it uses a lot of the training code. what it be better to include that in `train.py` and have additional argument to how frequent the plots are being generated?

I agree that evaluate.py does introduce code repetition but I still think it is useful to include it.

Its a lightweight version of train.py that isn't concerned with updating weights. And separating it from train.py, rather than integrating its logic with conditionals in there I think makes things more reader-friendly for newcomers to the codebase.

If we were to integrate this to train.py it isn't just plotting that will be considered but we'd need to have a option that runs a pretrained model on eval mode (so it doesn't update gradients and is more efficient), and also block saving of models and limit the epochs. Which again, isn't too hard to do - but I think will mess up clarity in train.py.

Also I've incorporated a plotting functionality for train.py already. It will output plots to 'src/out_train' whereas evaluate.py outputs plots to 'src/out_eval'

**Comment by RussellSB** [comment]
> > ## Preprocessing Discussion
> > In evaluate.py for the feats dict, I named the keys with respect to their conversion A2B, B2A. Seems to be a convention in style transfer code to use 'A2B' notation. Should we consider maybe changing the convention of how spkr directories are named in the database? Instead, if we were to name them spkr_A and spkr_B it could be more consistent. And we can iterate over key strings, instead of key int values in preprocessing.py.
> > Not only would this make things more consistent with "A2B notation" but say if the user were to name directories "spkr_4" or "spkr_7" (corresponding to ids from flickr_audio), it would string sort and still take spkr_4 as A and spkr_7 as B. We can then have n_spkrs controlled by the number of speaker directories in the dataset. Though for now, we'll note that it only works with two for the time being.
> 
> This seems to be a good idea; however, what do you think about having more than two speakers? we will need to use something else other than A B naming convention in that case. I will leave changing the naming to you if you want to do it or not.
> 
> > So for preprocessing maybe:
> > 
> > 1. Read number spkr dirs in dataset
> > 2. assert num_dirs == 2, 'Currently only two speakers accepted'
> 
> I like this assertion, we need to have it
> 
> > Would also need to reflect such changes in data_proc.py

Ok cool re naming conversion and naming them wrt wav dir names Ill keep this in mind, but Ill prioritise it for after test/eval/train split and inference.py

File: src/evaluate.py (hunk 4 above)
Comment (RussellSB) [reply src/evaluate.py]: Good idea, just added. Also like plot_interval added wav_interval to control how often it saves it. Also added an option to disable plot/wav saving when interval is set to -1.

File: src/preprocess.py [hunk 5]
```diff
@@ -14,7 +14,7 @@
 opt = parser.parse_args()
 print(opt)
 
-assert opt == 2, 'Currently only two speakers are supported'
+assert opt != 2, 'Currently only two speakers are supported'
```
Comment (ebadawy) [inline src/preprocess.py:17]: The first one was right, we need the number of speakers to be 2. it should also be `opt.n_spkrs`

Comment (RussellSB) [reply src/preprocess.py:17]: Right! Super silly mistake from my end. Sorry about that, getting it fixed now
